src/main/java/com/google/devtools/build/lib/rules/cpp/LibrariesToLinkCollector.java

```diff
@@ -236,6 +141,273 @@ public NestedSet<String> getRuntimeLibrarySearchDirectories() {
+    //
+    // For cases 1, 3, 5, 7, we need to compute the relative path from the output artifact to
+    // toolchain repo's bin directory. For 2 and 6, we walk down into `file.runfiles/<toolchain
+    // repo>`. For 4 and 8, we need to compute the relative path from the output runfile to
+    // <toolchain repo> under runfiles.
+    ImmutableList.Builder<String> solibParents = ImmutableList.builder();
+
+    // Cases 1, 3, 5, 7
+    PathFragment toolchainBin =
+        toolchainLibrariesSolibDir.getParentDirectory(); // relative to execroot
+    Preconditions.checkNotNull(toolchainBin);
+    PathFragment binaryOrigin = output.getExecPath().getParentDirectory(); // relative to execroot
+    Preconditions.checkNotNull(binaryOrigin);
+    solibParents.add(
+      getRelative(binaryOrigin, toolchainBin).getPathString() + PathFragment.SEPARATOR_CHAR);
+
+    // Cases 2 and 6
+    String toolchainRunfilesRepoName =
+        getRunfilesRepoName(ccToolchainProvider.getCcToolchainLabel().getRepository());
+    solibParents.add(
+        PATH_JOINER.join(output.getFilename() + ".runfiles", toolchainRunfilesRepoName)
+            + PathFragment.SEPARATOR_CHAR);
+
+    // Cases 4 and 8
+    String binaryRepoName = getRunfilesRepoName(
+        Preconditions.checkNotNull(output.getOwnerLabel()).getRepository());
+    toolchainBin = PathFragment.create(toolchainRunfilesRepoName); // relative to runfiles
+    binaryOrigin =
+        PathFragment.create(binaryRepoName)
+            .getRelative(output.getRepositoryRelativePath())
+            .getParentDirectory();
+    Preconditions.checkNotNull(binaryOrigin);
+    solibParents.add(
+      getRelative(binaryOrigin, toolchainBin).getPathString() + PathFragment.SEPARATOR_CHAR);
+
+    return solibParents.build();
+  }
+
+  private String getRunfilesRepoName(RepositoryName repo) {
+    if (repo.isMain()) {
+      return workspaceName;
+    }
+    return repo.getName();
+  }
+
+  /**
+   * Returns the relative {@link PathFragment} from "from" to "to".
+   *
+   * <p>Example 1: <code>
+   * getRelative({@link PathFragment}.create("foo"), {@link PathFragment}.create("foo/bar/wiz"))
+   * </code> returns <code>"bar/wiz"</code>.
+   *
+   * <p>Example 2: <code>
+   * getRelative({@link PathFragment}.create("foo/bar/wiz"), {@link PathFragment}.create("foo/wiz"))
+   * </code> returns <code>"../../wiz"</code>.
+   *
+   * <p>The following requirements / assumptions are made: 1) paths must be both absolute or both
+   * relative; 2) for absolute windows paths, they must be under the same drive letter; 3) when
+   * paths are relative, they are assumed to be relative to the same location; 4) when the {@code
+   * from} path starts with {@code ..} prefixes, the prefix length must not exceed {@code ..}
+   * prefixes of the {@code to} path.
+   */
+  static PathFragment getRelative(PathFragment from, PathFragment to) {
+    if (from.isAbsolute() != to.isAbsolute()) {
```

Comment:
I wonder whether this complex function is really needed. Are we every passing absolute paths into it?

I came up with the following simpler implementation that passes the two toolchain tests:
```java
List<String> fromSegments = from.splitToListOfSegments();
List<String> toSegments = to.splitToListOfSegments();

int firstMismatch = 0;
while (firstMismatch < fromSegments.size()
    && firstMismatch < toSegments.size()
    && OS.equals(fromSegments.get(firstMismatch), toSegments.get(firstMismatch))) {
  firstMismatch++;
}

return PathFragment.create("../".repeat(fromSegments.size() - firstMismatch))
    .getRelative(to.subFragment(firstMismatch, toSegments.size()));
```

Reply:
Yep we can simplify it a bit by accepting relative paths only (as we are not passing absolute paths to it).

Some necessary checks are still needed: 1) enforcing relative path and 2) check for extra `..` segments in `from`.